app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java

```diff
@@ -0,0 +1,552 @@
+    // onLoginClicked() - attempts to get tokens for the API, if it succeeds calls the API & updates UI
+    // resetPassword() - resets the user password then kicks off the login flow.
+    // onEditProfileClicked() - pops dialogue for user to edit their profiel then gets fresh tokens.
+    // onSignOutClicked() - Signs account out of the app & updates UI
+    // callApi() - called on successful token acquisition which makes an HTTP request to our API
+    //
+
+    /* Handles the redirect from the System Browser */
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        sampleApp.handleInteractiveRequestRedirect(requestCode, resultCode, data);
+    }
+
+    private void silentRequest(boolean forceRefresh) {
+        /* Attempt to get a user and get a token silently
+         * If this fails we will do an interactive request
+         */
+        List<IAccount> accounts = null;
+
+        try {
+            accounts = sampleApp.getAccounts();
+
+            if (accounts != null && accounts.size() == 1) {
+                /* We have 1 account */
+                sampleApp.acquireTokenSilentAsync(
+                        B2C_SCOPES,
+                        accounts.get(0),
+                        SISU_POLICY,
+                        forceRefresh,
+                        getAuthSilentCallback());
+            } else {
+                /* We have no account or 2 or more accounts */
+            }
+        } catch (IndexOutOfBoundsException e) {
+            Log.d(TAG, "Account at this position does not exist: " + e.toString());
+        }
+
+    }
+
+    /* Use MSAL to acquireToken for the end-user
+     * Callback will call api w/ access token & update UI
+     */
+    private void onLoginClicked() {
+        if (curDomainHint != null) {
+            List idProvider = new ArrayList<Pair<String, String>>();
+            idProvider.add(new Pair<String, String>("domain_hint", curDomainHint));
+            sampleApp.acquireToken(
+                    getActivity(),
+                    B2C_SCOPES,
+                    (String) null,
+                    UiBehavior.SELECT_ACCOUNT,
+                    idProvider,
+                    getAuthInteractiveCallback());
+        } else {
+            sampleApp.acquireToken(
+                    getActivity(),
+                    B2C_SCOPES,
+                    getAuthInteractiveCallback());
+        }
+    }
+
+    /* Use MSAL to reset the users password
+     * Callback should call a SiSu policy I think
+     */
+    private void resetPassword() {
+        sampleApp.acquireToken(
+                getActivity(),
+                B2C_RESET_SCOPES,
+                RESET_PASSWORD_POLICY,
+                getResetCallback());
+    }
+
+    /* Use MSAL to edit their profile
+     * Callback will land them on authenticated page.
+     */
+    private void onEditProfileClicked() {
+        if (curDomainHint != null) {
```

Comment:
Can be simplied like below
`

        List<Pair<String, String>> idProvider = null;

        if(!TextUtils.isEmpty(curDomainHint)){
            idProvider = new ArrayList<>();
            idProvider.add(new Pair<String, String>("domain_hint", curDomainHint));
        }

        sampleApp.acquireToken(
                getActivity(),
                B2C_EDIT_SCOPES,
                (String) null,
                UiBehavior.SELECT_ACCOUNT,
                idProvider,
                null,
                EDIT_PROFILE_POLICY,
                getEditCallback());
`